### ISSUE TITLE
Polymer vs the ecmascript package

### ISSUE THREAD

**Issue by yorrd** [meteor/meteor#9717]
So I'm trying to use Polymer webcomponents (using their [polyfill](https://github.com/webcomponents/webcomponentsjs)) with Meteor's ecmascript plugin. Why is this an issue? Ecmascript compiles the constructor away and webcomponents want to be accessed via constructor. 

This is the error being thrown: `Uncaught TypeError: Failed to construct 'HTMLElement': Please use the 'new' operator, this DOM object constructor cannot be called as a function.` because the following code gets called with `new`: 

```
    function PropertiesChanged() {
      var _this;

      _this = _superClass.call(this) || this;
```

... which obviously is a function, not a class. In the non-compiled source code, there is a class. 

As far as I can see I have three options:

- compile the polymer library with the same babel config as meteor (it tries to use the latest supported features, Meteor apparently doesn't)
- write a polyfill for a polyfill (nope.)
- ask for help because both other options really suck

---

**`<rant>`** This is already a pain because I have to copy the Polymer library from `node_modules` to `imports`, otherwise it [won't get compiled](https://github.com/meteor/meteor-feature-requests/issues/6) and browsers don't support es6 and import statements yet which Polymer - like many libraries - uses. Therefore however, I have no issue compiling the library. When node_modules get compiled, this will get even worse, because I can't even manually compile Polymer without copying it over again. **`</rant>`**

**Comment by yorrd** [comment]
@hwillson sure. Have a look here: https://github.com/yorrd/base

A few remarks:

- don't test on Firefox, Polymer runs with a polyfill there and you won't be able to reproduce it. On Chrome there is no need for a polyfill and therefore the issue will arise
- the main file is https://github.com/yorrd/base/blob/master/imports/ui/components/polymer-3-test.ts
- I've copied the polymer libraries into imports/ui/components/node_links so they will be compiled, therefore the import statements import from there, not from node_modules
- the polymer-compiler package only adds the head and body contents (like static-html does as well), nothing fancy
- this is our internal base project (the stripped down version anyways). Please forgive me if I missed something which is not necessary to the reproduction

with this in mind, you should see the error in the chrome console which I described above.

**Comment by benjamn** [comment]
What you're bumping into here is a general problem with compiled ECMAScript classes. Specifically, there is no way for a compiled subclass to invoke the `constructor` of a native (not compiled) superclass, because the only way to do that correctly is to call `super(...)`, which only works in native classes. The old trick of calling `SuperClass.call(this, ...)` no longer works because TC39 decided to make `class` constructors throw if you try to call them as functions rather than using `new`. This decision has been a major problem for the Babel ecosystem, which includes Meteor.

What this means is that any library that exports a native `class` that is intended to be subclassed cannot be consumed by codebases that compile their `class`es down to ES5 functions. While the Polymer team may have had their reasons for this decision, they have effectively made their library useless to most Babel-compiled projects, except for those that do not bother supporting older browsers at all. If they had shipped the Polymer element classes as compiled ES5 functions, then both native `class`es and compiled ES5 constructor function-based classes could subclass them, but apparently that wasn't something they cared about.

# However! 👇👀🎉🐙🔥

Meteor 1.6.2 will **solve** this problem, by bundling your client-side JavaScript in a radically different way from any previous version of Meteor, and indeed most other JavaScript frameworks in use today!

Instead of building a single JS bundle using Babel compiler options appropriate for the oldest supported browsers, Meteor 1.6.2 will build two bundles: a _legacy_ bundle that works the same way as ever, and a _modern_ bundle that targets only recent evergreen browsers, which roughly corresponds to browser with native `async`/`await` support. Amazingly, this represents more than 80% of browsers in use today, and that percentage is only going to keep growing.

In particular, every browser that qualifies as "modern" has native support for ECMAScript `class` syntax, so there's no need to compile that syntax for the modern bundle, so Polymer element subclassing should _just work_ in modern browsers. Best of all, the selection of which bundle to send to which client is completely automatic

In other words, you have one more option, which is to start using the Meteor 1.6.2 betas. Something tells me that won't bother you too much, since you're already an early adopter of the Polymer stuff. 😉

**tl;dr** Please try running
```
meteor update --release 1.6.2-beta.13
meteor npm install @babel/runtime@latest
meteor run
```
in your app. As with any update, there may be some small issues to fix, but I think you're going to love the final result.

Once you start using Meteor 1.6.2, I hope you'll [keep up with the beta versions](https://github.com/meteor/meteor/pull/9559) and release candidates, and let us know if you hit any more snags so that we can fix them before the final release!

**Comment by yorrd** [comment]
@benjamn Thank you. I love you.

First of all: makes sense with the constructor stuff. At the moment, I'm just using a custom babel compiler to compile the `node_modules/@polymer/` directory. 

Question about the new compilation in 1.6.2: the old bundle will have the same problem won't it? And another question: while `class` is supported by all modern browsers, es6 module importing is not as far as I know. That's the second major hurdle which we currently get around by compiling it down to commonjs imports with that custom babel compiler as well. 

You're completely right, we love cutting edge and save a lot of money off of it for our customers, but we do have to support some "non-modern" browsers. I really look forward to testing the betas!

**Comment by benjamn** [comment]
Yes, the legacy bundle will still need `node_modules/@polymer` to be compiled, which is not currently something that Meteor does. One of the goals of Meteor 1.6.2 is to start compiling `node_modules` for the client bundle—but only for the legacy bundle, since modern browsers can probably handle anything published to npm. This would solve a lot of problems, since more and more npm packages are being published with `class`, `const`, `let`, `Symbol`, and friends.

**Comment by benjamn** [comment]
Here's how we're planning to improve this functionality in Meteor 1.6.2: #9771 

Specifically, you can `meteor npm install @polymer/polymer` as usual, and then create a symbolic link to `node_modules/@polymer/polymer` somewhere in your application, which will cause the Polymer code to be compiled appropriately for each platform (`web.browser`, `web.browser.legacy`, and `os.*`). When you `import stuff from @polymer/polymer` you'll get the compiled code, rather than the raw code installed from npm.

**Comment by yorrd** [comment]
@benjamn I don't think I'm doing this correctly. If I have the ecmascript package installed and I link the libraries to a subdirectory of the `imports/` module, it should compile to es5, right? However, when I load the page, it gives me an error about import statements which are not compiled away. Am I missing something obvious?

**Comment by benjamn** [comment]
@yorrd What does your directory structure end up looking like? Have you run `meteor update --release 1.6.2-beta.14`?

**Comment by yorrd** [comment]
@benjamn yes I have updated. The directory structure is 

```
imports
-> compile
-> -> polymer (link)
-> -> polymer component 1 (link)
-> -> polymer component 2 (link)
-> -> other lib (link)
-> ui... the usual
node_modules
-> @polymer
-> -> polymer
-> -> polymer comp 1
-> -> polymer comp 2
```

Does that help?

**Comment by benjamn** [comment]
That looks reasonable, so I don't think you're missing something obvious. If there's a small reproduction you could provide, we can take a look at it. Thanks for your patience!

**Comment by yorrd** [comment]
@benjamn had some issues. This is the one I already linked once with the updated stuff. Not sure if symlinks get transferred over git, you might have to add them again: https://github.com/yorrd/base

**Comment by yorrd** [comment]
Any news on this? Can I help in any way?

**Comment by Dev1an** [comment]
@benjamn Is it normal that I get `false` for `Meteor.isModern` on my iPad with iOS 11.3.1?

**Comment by benjamn** [comment]
@Dev1an I think you've identified the problem! Fortunately, this system is designed to allow fixing bugs like this simply by republishing the affected packages (no new Meteor release required). And of course the system is behaving as designed by falling back to the legacy bundle when we can't be sure a browser is modern, even if the reason for that uncertainty is a bug that should be fixed.

**Comment by benjamn** [comment]
@yorrd Finally looking at your repo, and I think [the symbolic link](https://github.com/yorrd/base/blob/master/imports/compile/polymer) is slightly off. Don't you need `../../` instead of just `../`?

**Comment by yorrd** [comment]
@benjamn I'm not sure how I missed this. There was another issue with babel which probably confused me. Thanks!
